File: src/main/java/dark/leech/text/plugin/js/api/JSElement.java
package dark.leech.text.plugin.js.api;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import dark.leech.text.action.Log;

/**
 * Rhino wrapper for JSoup Element. Simple Java class - public methods automatically exposed to
 * JavaScript.
 */
public class JSElement {
    private final Element element;

    public JSElement(Element element) {
        this.element = element;
    }

    /** Select child elements using CSS selector. Returns JSElements collection for chaining. */
    public JSElements select(String selector) {
        try {
            if (element == null) {
                Log.add(
                        "WARNING: JSElement.select() called on null element with selector: "
                                + selector);
                return new JSElements(new Elements());
            }
            if (selector == null || selector.isEmpty()) {
                return new JSElements(new Elements());
            }
            Elements results = element.select(selector);
            return new JSElements(results);
        } catch (Exception e) {
            Log.add(
                    "ERROR: JSElement.select() failed for selector '"
                            + selector
                            + "': "
                            + e.getMessage());
            return new JSElements(new Elements());
        }
    }

    /** Get element text content. */
    public String text() {
        return element != null ? element.text() : "";
    }

    /** Get element own text (excluding children). */
    public String ownText() {
        return element != null ? element.ownText() : "";
    }

    /** Get element HTML content. */
    public String html() {
        return element != null ? element.html() : "";
    }

    /** Get element outer HTML. */
    public String outerHtml() {
        return element != null ? element.outerHtml() : "";
    }

    /** Get attribute value. */
    public String attr(String key) {
        if (key == null || key.isEmpty()) {
            return "";
        }
        return element != null ? element.attr(key) : "";
    }

    /** Set attribute value. */
    public JSElement attr(String key, String value) {
        if (element != null && key != null && !key.isEmpty()) {
            element.attr(key, value);
        }
        return this;
    }

    /** Check if element has class. */
    public boolean hasClass(String className) {
        if (className == null || className.isEmpty()) {
            return false;
        }
        return element != null && element.hasClass(className);
    }

    /** Get element tag name. */
    public String tagName() {
        return element != null ? element.tagName() : "";
    }

    /** Get element id. */
    public String id() {
        return element != null ? element.id() : "";
    }

    /** Get element class name. */
    public String className() {
        return element != null ? element.className() : "";
    }

    /** Get first child element. */
    public JSElement first() {
        if (element == null) {
            return null;
        }
        Element first = element.firstElementChild();
        return first != null ? new JSElement(first) : null;
    }

    /** Get last child element. */
    public JSElement last() {
        if (element == null) {
            return null;
        }
        Element last = element.lastElementChild();
        return last != null ? new JSElement(last) : null;
    }

    /** Get child elements. */
    public JSElements children() {
        if (element == null) {
            return new JSElements(new Elements());
        }
        return new JSElements(element.children());
    }

    /** Get parent element. */
    public JSElement parent() {
        if (element == null) {
            return null;
        }
        Element parent = element.parent();
        return parent != null ? new JSElement(parent) : null;
    }

    /** Get next sibling element. */
    public JSElement next() {
        if (element == null) {
            return null;
        }
        Element next = element.nextElementSibling();
        return next != null ? new JSElement(next) : null;
    }

    /** Get previous sibling element. */
    public JSElement prev() {
        if (element == null) {
            return null;
        }
        Element prev = element.previousElementSibling();
        return prev != null ? new JSElement(prev) : null;
    }

    /** Check if element is empty (no children). */
    public boolean isEmpty() {
        return element == null || element.children().isEmpty();
    }

    /** Get number of child elements. */
    public int childrenSize() {
        return element != null ? element.children().size() : 0;
    }

    /** Remove element. */
    public void remove() {
        if (element != null) {
            element.remove();
        }
    }

    /** Remove elements matching selector. */
    public JSElement remove(String selector) {
        if (element != null && selector != null && !selector.isEmpty()) {
            element.select(selector).remove();
        }
        return this;
    }

    /** Add class. */
    public JSElement addClass(String className) {
        if (element != null && className != null && !className.isEmpty()) {
            element.addClass(className);
        }
        return this;
    }

    /** Remove class. */
    public JSElement removeClass(String className) {
        if (element != null && className != null && !className.isEmpty()) {
            element.removeClass(className);
        }
        return this;
    }

    /** Toggle class. */
    public void toggleClass(String className) {
        if (element != null && className != null && !className.isEmpty()) {
            element.toggleClass(className);
        }
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/api/JSElement.kt
package dev.haipham22.leechtext.plugin.js.api

import com.fleeksoft.ksoup.select.Elements

/**
 * Wrapper ksoup Element cho Rhino (port từ JSElement.java; P4: jsoup → ksoup).
 */
@Suppress("TooManyFunctions") // API surface cho plugin JS — số hàm mirror đúng API vBook, không tách
class JSElement(
    val element: com.fleeksoft.ksoup.nodes.Element?,
) {
    fun select(selector: String?): JSElements {
        return try {
            if (element == null) {
                return JSElements(Elements(arrayListOf()))
            }
            if (selector.isNullOrEmpty()) {
                return JSElements(Elements(arrayListOf()))
            }
            JSElements(element.select(selector))
        } catch (e: Exception) {
            JSElements(Elements(arrayListOf()))
        }
    }

    fun text(): String = element?.text() ?: ""

    fun ownText(): String = element?.ownText() ?: ""

    fun html(): String = element?.html() ?: ""

    fun outerHtml(): String = element?.outerHtml() ?: ""

    fun attr(key: String?): String {
        if (key.isNullOrEmpty()) return ""
        return element?.attr(key) ?: ""
    }

    fun attr(
        key: String?,
        value: String?,
    ): JSElement {
        if (element != null && !key.isNullOrEmpty()) element.attr(key, value ?: "")
        return this
    }

    fun hasClass(className: String?): Boolean {
        if (className.isNullOrEmpty()) return false
        return element?.hasClass(className) == true
    }

    fun tagName(): String = element?.tagName() ?: ""

    fun id(): String = element?.id() ?: ""

    fun className(): String = element?.className() ?: ""

    fun first(): JSElement? = element?.firstElementChild()?.let { JSElement(it) }

    fun last(): JSElement? = element?.lastElementChild()?.let { JSElement(it) }

    fun children(): JSElements = if (element == null) {
        JSElements(Elements(arrayListOf()))
    } else {
        JSElements(Elements(element.children()))
    }

    fun parent(): JSElement? = element?.parent()?.let { JSElement(it) }

    fun next(): JSElement? = element?.nextElementSibling()?.let { JSElement(it) }

    fun prev(): JSElement? = element?.previousElementSibling()?.let { JSElement(it) }

    fun isEmpty(): Boolean = element == null || element.children().isEmpty()

    fun childrenSize(): Int = element?.children()?.size ?: 0

    fun remove() {
        element?.remove()
    }

    fun remove(selector: String?): JSElement {
        // ksoup Elements không có remove() no-arg như jsoup — detach từng element
        if (element != null && !selector.isNullOrEmpty()) element.select(selector).forEach { it.remove() }
        return this
    }

    fun addClass(className: String?): JSElement {
        if (element != null && !className.isNullOrEmpty()) element.addClass(className)
        return this
    }

    fun removeClass(className: String?): JSElement {
        if (element != null && !className.isNullOrEmpty()) element.removeClass(className)
        return this
    }

    fun toggleClass(className: String?) {
        if (element != null && !className.isNullOrEmpty()) element.toggleClass(className)
    }
}
